monai/utils/arrayutils.py

```diff
@@ -139,9 +140,121 @@ def resize_center(img, *resize_dims, fill_value=0):
+    """
+    ndim = len(dims)
+
+    try:
+        # if a single value was given as patch size, treat this as the size of the patch over all dimensions
+        single_patch_size = int(patch_size)
+        patch_size = (single_patch_size,) * ndim
+    except TypeError:  # raised if the patch size is multiple values
+        # ensure patch size is at least as long as number of dimensions
+        patch_size = ensure_tuple_size(patch_size, ndim)
+
+    # ensure patch size dimensions are not larger than image dimension, if a dimension is None or 0 use whole dimension
+    return tuple(min(ms, ps or ms) for ms, ps in zip(dims, patch_size))
+
+
+def get_random_patch(dims, patch_size):
+    """
+    Returns a tuple of slices to define a random patch in an array of shape `dims` with size `patch_size` or the as
+    close to it as possible within the given dimension. It is expected that `patch_size` is a valid patch for a source
+    of shape `dims` as returned by `get_valid_patch_size`.
+
+    Args:
+        dims (tuple of int): shape of source array
+        patch_size (tuple of int): shape of patch size to generate
+
+    Returns:
+        (tuple of slice): a tuple of slice objects defining the patch
+    """
+
+    # choose the minimal corner of the patch
+    min_corner = tuple(np.random.randint(0, ms - ps) if ms > ps else 0 for ms, ps in zip(dims, patch_size))
+
+    # create the slices for each dimension which define the patch in the source array
+    return tuple(slice(mc, mc + ps) for mc, ps in zip(min_corner, patch_size))
+
+
+def iter_patch_slices(dims, patch_size, start_pos=()):
+    """
+    Yield successive tuples of slices defining patches of size `patch_size` from an array of dimensions `dims`. The 
+    iteration starts from position `start_pos` in the array, or starting at the origin if this isn't provided. Each 
+    patch is chosen in a contiguous grid using a first dimension as least significant ordering.
+
+    Args:
+        dims (tuple of int): dimensions of array to iterate over
+        patch_size (tuple of int or None): size of patches to generate slices for, 0 or None selects whole dimension
+        start_pos (tuple of it, optional): starting position in the array, default is 0 for each dimension
+
+    Yields:
+        Tuples of slice objects defining each patch
+    """
+
+    # ensure patchSize and startPos are the right length
+    ndim = len(dims)
+    patch_size = get_valid_patch_size(dims, patch_size)
+    start_pos = ensure_tuple_size(start_pos, ndim)
+
+    # collect the ranges to step over each dimension
+    ranges = tuple(starmap(range, zip(start_pos, dims, patch_size)))
+
+    # choose patches by applying product to the ranges
+    for position in product(*ranges[::-1]):  # reverse ranges order to iterate in index order
+        yield tuple(slice(s, s + p) for s, p in zip(position[::-1], patch_size))
+
+
+def iter_patch(arr, patch_size, start_pos=(), copy_back=True, pad_mode="wrap", **pad_opts):
```

Comment:
I think this needs to be extended to support dense sampling grid (overlapping patches)  so that it could be used for #39

Reply:
Yes I was thinking this too for other tasks.